### PR TITLE
Fix a segfault in rest_collection

### PR DESCRIPTION
**Pull request by FinchPowers**
It occurrs when the object is destroyed before threads have finished
running. I added a counter and a condition variable to have the dtor
wait for them before dtoring.

### CODE REVIEW

File: rest/rest_collection_impl.h [hunk 1]
```diff
@@ -82,6 +82,10 @@ struct RestCollection<Key, Value>::Impl {
         protected by the RCU lock.
     */
     MutateMutex mutateMutex; // not mutable on purpose
+
+    int runningThreads;
```
Comment (FinchPowers) [inline rest/rest_collection_impl.h]: I added a counter, but my knowledge of that class is insufficient to know whether there can be more than one or not.

File: rest/rest_collection_impl.h [hunk 2]
```diff
@@ -143,6 +147,13 @@ void
 RestCollection<Key, Value>::
 clear()
 {
+    {
+        std::unique_lock<std::mutex> lock(impl->runningThreadsMutex);
+        while (impl->runningThreads != 0) {
+            std::cerr << "running threads: " << impl->runningThreads << std::endl;
```
Comment (FinchPowers) [inline rest/rest_collection_impl.h]: Debug output. Can be removed.

**Comment by FinchPowers** [comment]
One bug with that fix: If the thread crashes before the counter can be decremented, the program will hang there.

**Comment by guyd** [comment]
We are getting closer to a fix.
Deferring to Jeremy to confirm that such an approach can work.

**Comment by jeremybarnes** [comment]
Bravo for the diagnosis @FinchPowers .  It doesn't entirely fix the problem (I have a testcase that still triggers an error, although less frequently with your patch).  I'm working on a better fix.

**Comment by jeremybarnes** [comment]
@FinchPowers FYI.

**Comment by guyd** [comment]
Looking forward to see that merged into master.

**Pull request merged**